### ISSUE TITLE
Unable to resume wandb reporting when resuming training

### ISSUE THREAD

**Issue by JairajJangle** [ultralytics/ultralytics#22019]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

I am setting up wandb like this, before I call `model.train(...)` for **both**, new train as well as for resume train:
```python
!pip install -U wandb
!yolo settings wandb=True
wandb.login(key="XXX")
```

---

For a new train, I mention the name of the project and the run name like so:
```python
model = YOLO('my-yolo-model.pt')

results = model.train(
    data='/path/to/my/dataset/data.yaml',  
    ...
    project=project,
    name=run_name
)
```
At the end of each epoch, the metrics get logged over to wandb for a new train.

---

However, when I resume the training, wandb doesn't seem to work at all. Here is how I resume the training:
```python
model = YOLO("path-to-my-yolo-run/last.pt")
results = model.train(resume=True)
```

I also tried adding my own callback function to report the metrics to wandb like this, before I resume training:
```python
model = YOLO("path-to-my-yolo-run/last.pt")

wandb.init(
    project=project,
    id="my-wand-b-run-id",
    resume="must"  # This ensures it resumes the existing run
)

# Add a callback to manually log metrics - because ultralytics wandb logging doesn't seem to work after resuming training
def log_metrics(trainer):
    if hasattr(trainer, 'metrics') and trainer.metrics:
        metrics_dict = {}
        for key, value in trainer.metrics.items():
            if isinstance(value, (int, float)):
                metrics_dict[key] = value

        if metrics_dict:
            wandb.log(metrics_dict, step=trainer.epoch)
# Add the callback
model.add_callback('on_fit_epoch_end', log_metrics)

results = model.train(resume=True)
```

This partially works; it reports metrics & validation losses, but it does not report training losses. I couldn't seem to find a way to report training loss at all. 

This seems to be a bug because I remember wandb reporting used to resume seamlessly on `ultralytics==8.0.x` version.

### Environment

```
Ultralytics 8.3.197 🚀 Python-3.11.7 torch-2.1.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48670MiB)
Setup complete ✅ (8 CPUs, 44.1 GB RAM, 107.1/244.2 GB disk)

OS                     Linux-5.19.0-45-generic-x86_64-with-glibc2.35
Environment            Docker
Python                 3.11.7
Install                pip
Path                   /usr/local/lib/python3.11/dist-packages/ultralytics
RAM                    44.08 GB
Disk                   107.1/244.2 GB
CPU                    Intel Xeon Gold 5315Y CPU @ 3.20GHz
CPU count              8
GPU                    NVIDIA RTX A6000, 48670MiB
GPU count              1
CUDA                   12.1

numpy                  ✅ 1.26.3>=1.23.0
matplotlib             ✅ 3.7.3>=3.3.0
opencv-python          ✅ 4.8.0.76>=4.6.0
pillow                 ✅ 9.5.0>=7.1.2
pyyaml                 ✅ 5.4.1>=5.3.1
requests               ✅ 2.31.0>=2.23.0
scipy                  ✅ 1.11.2>=1.4.1
torch                  ✅ 2.1.1+cu121>=1.8.0
torch                  ✅ 2.1.1+cu121!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.16.1+cu121>=0.9.0
psutil                 ✅ 5.9.8
polars                 ✅ 1.33.1
ultralytics-thop       ✅ 2.0.17>=2.0.0
```

### Minimal Reproducible Example

```python
!pip install -U wandb
!yolo settings wandb=True
wandb.login(key="XXX")

# For new training:
model = YOLO('my-yolo-model.pt')

results = model.train(
    data='/path/to/my/dataset/data.yaml',  
    ...
    project=project,
    name=run_name
)

# For resume training:
model = YOLO("path-to-my-yolo-run/last.pt")
results = model.train(resume=True)
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @JairajJangle, thanks for reporting this and for using Ultralytics 🚀 This is an automated assistant response — an Ultralytics engineer will review and assist you here shortly.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. For wandb resume issues specifically, please include:
-Exact commands and code used to start the original run and to resume it
-Full console logs from both sessions (including any wandb messages)
-The exact run directory path used for resume and the `train*.yaml` from that run
-Any custom callbacks you added and where you registered them

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
-**Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
-**Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
-**Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
-**Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

💡 To help us narrow this down further, please also confirm:
-Your exact `ultralytics` version and whether the issue persists after `pip install -U ultralytics`
-Whether the same behavior occurs when resuming from `last.pt` vs `best.pt`
-Whether wandb logs appear in the run directory (e.g., `wandb/latest-run`) after resume

Thanks again for the detailed report — once we have the above details or a small repro script, we’ll investigate promptly.

**Comment by JairajJangle** [comment]
Replying to @UltralyticsAssistant :

> -Your exact ultralytics version and whether the issue persists after pip install -U ultralytics

`8.3.197` (already the latest version).


> -Whether the same behavior occurs when resuming from last.pt vs best.pt

No, I didn't check this. I don't think it's appropriate to use `best.pt` to resume training. I use `last.pt` as per the documentation over here: https://docs.ultralytics.com/modes/train/#resuming-interrupted-trainings


> -Whether wandb logs appear in the run directory (e.g., wandb/latest-run) after resume

Yes, wandb logs appear in the run directory: `wandb/latest-run` after resume.

**Comment by Y-T-G** [comment]
You can remove these:
```
# Add a callback to manually log metrics - because ultralytics wandb logging doesn't seem to work after resuming training
def log_metrics(trainer):
    if hasattr(trainer, 'metrics') and trainer.metrics:
        metrics_dict = {}
        for key, value in trainer.metrics.items():
            if isinstance(value, (int, float)):
                metrics_dict[key] = value

        if metrics_dict:
            wandb.log(metrics_dict, step=trainer.epoch)
# Add the callback
model.add_callback('on_fit_epoch_end', log_metrics)
```

and just keep the `wandb.init()` before you start training

**Comment by Y-T-G** [comment]
Also, I don't think this is a bug. Ultralytics doesn't seem to have ever had any inherent resume feature for W&B because the run ID was never saved, W&B runs can't be resumed without run ID.

**Comment by JairajJangle** [comment]
Thanks for the info @Y-T-G, I will try it out.

**Comment by JairajJangle** [comment]
Doing just `wandb.init()` does not work, I have tried in the past as well. None of the metrics get reported over to wandb after resuming the run. This is my whole code for resume training:

```python
!pip install ultralytics

from ultralytics import YOLO

!nvidia-smi

!pip install -U wandb
!yolo settings wandb=True
import wandb

wandb.login(key="XXX")

project = "my_project_name_here"
run_name = "run_name_here"

model = YOLO("./path-to-my-yolo-run/weights/last.pt") 

wandb.init(
    project=project,
    id="wandb_run_id",  # Use the same ID from interrupted run
    resume="must"  # This ensures it resumes the existing run
)

# Commented this as suggested
# # Add a callback to manually log metrics - because ultralytics wandb logging doesn't seem to work after resuming training
# def log_metrics(trainer):
#    if hasattr(trainer, 'metrics') and trainer.metrics:
#        metrics_dict = {}
#        for key, value in trainer.metrics.items():
#            if isinstance(value, (int, float)):
#                metrics_dict[key] = value
# 
#        if metrics_dict:
#            wandb.log(metrics_dict, step=trainer.epoch + 1)
# # Add the callback
#  model.add_callback('on_fit_epoch_end', log_metrics)

# Resume training
results = model.train(resume=True)
```

Here is the output of `wandb.init(...)`:

<img width="734" height="636" alt="Image" src="https://github.com/user-attachments/assets/31f656d1-9a51-4ce3-880e-3aafc644b4a0" />

---

For now, I am going to revert to using my custom callback (as in the commented code), at least it used to report all the metrics except for train losses & lr pg0/1/2

**Comment by Y-T-G** [comment]
You need to import Ultralytics after running `!yolo settings wandb=True`, not before. Otherwise, the settings wouldn't take effect.

**Comment by JairajJangle** [comment]
I don't see any specifics of this in the documentation. Regardless, I have made the changes, as I expected, none of the metrics/losses get reported over to wandb. 

```python
!pip install ultralytics

!nvidia-smi

!pip install -U wandb
!yolo settings wandb=True                 # <----------------------- HERE IS THE SETTING OF WANDB TO TRUE
import wandb

wandb.login(key="XXX")

project = "my_project_name_here"
run_name = "run_name_here"

from ultralytics import YOLO              # <----------------------- MOVED THIS HERE

model = YOLO("./path-to-my-yolo-run/weights/last.pt") 

wandb.init(
    project=project,
    id="wandb_run_id",  # Use the same ID from interrupted run
    resume="must"  # This ensures it resumes the existing run
)

# Commented this as suggested
# # Add a callback to manually log metrics - because ultralytics wandb logging doesn't seem to work after resuming training
# def log_metrics(trainer):
#    if hasattr(trainer, 'metrics') and trainer.metrics:
#        metrics_dict = {}
#        for key, value in trainer.metrics.items():
#            if isinstance(value, (int, float)):
#                metrics_dict[key] = value
# 
#        if metrics_dict:
#            wandb.log(metrics_dict, step=trainer.epoch + 1)
# # Add the callback
#  model.add_callback('on_fit_epoch_end', log_metrics)

# Resume training
results = model.train(resume=True)
```

---

Hence, this definitely seems like a BUG to me. Working and tested code snippets/pointers to official docs would be highly appreciated.

**Comment by Y-T-G** [comment]
It should work. In fact, I tested it before I replied to you earlier and it does work and resumes perfectly fine.

It seems like an issue with your environment. What environment are you using and why do you have to install Ultralytics and W&B every time? If resuming isn't working, then I would also assume that training from scratch also doesn't report any metrics to W&B for you.

**Comment by JairajJangle** [comment]
Hi @Y-T-G thank you for the continued follow-up responses. The issue might be with my setup. I created a clean project and re-ran everything from scratch, and it seems to work now. However, just to be sure, I'll keep a watch on it and probably close the issue in 1-2 days after I see that it's working every time. 

Meanwhile, I think it would be a good idea to document that the import statement `from ultralytics import YOLO` should come after setting wandb as True. 

---

Edit: Nope, something is wrong, it worked for the first time after resuming the run after doing everything from scratch. But wandb reporting stopped again after resuming the run thereafter. For reference, I am using Paperspace Gradient Notebook to do the training, hence, I am required to install `ultralytics` & `wandb` each time before I resuming the training.

**Comment by Y-T-G** [comment]
> Meanwhile, I think it would be a good idea to document that the import statement `from ultralytics import YOLO` should come after setting wandb as True.

This problem arises because you're mixing two different modes of usage, CLI and Python API, and also doing so in a notebook. You should either stick to CLI or stick to Python API, not mix both in the same session. To update settings using Python API, you can run:
```python
from ultralytics import SETTINGS
SETTINGS["wandb"] = True
```

**Comment by JairajJangle** [comment]
```python
from ultralytics import SETTINGS
SETTINGS["wandb"] = True
```

After doing this, it seems to work every time now. Probably, a good idea to document the precise conditions required to make wandb work as suggested in this conversation. I am open to raising a PR for the doc update changes. Let me know. 

---

Thanks @Y-T-G for the help!
